lua/docgen/renderer.lua

```diff
@@ -0,0 +1,174 @@
+local m = {}
+
+local Text = {}
+Text.__index = Text
+
+function Text:new()
+  return setmetatable({
+    state = nil,
+    paragraphes = {},
+    itemizes = {},
+    enumerates = {},
+    order = {}
+  }, Text)
+end
+
+local interpret_state = function(line)
+  if line == '' then
+    return 'newline'
+  elseif vim.startswith(vim.trim(line), '-') then
+    return 'itemize'
+  elseif vim.trim(line):match('^[0-9]') then
+    return 'enumerate'
+  else
+    return 'paragraph'
+  end
+end
+
+local states = {
+  ['paragraph'] = function(self, line) return self:add_to_paragraph(line) end,
+  ['newline'] = function(self) self:start_new_paragraph(); self:start_new_paragraph() end,
+  ['itemize'] = function(self, line) return self:add_to_itemize(line) end,
+  ['enumerate'] = function(self, line) return self:add_to_enumerate(line) end,
+}
+
+local append = function(l, r)
+  if #l == 0 then
+    return r
+  end
+  if l:sub(#l, #l) == ' ' then
+    return l .. r
+  else
+    return l .. ' ' .. r
+  end
+end
+
+function Text:handle_line(line)
+  local new_state = interpret_state(line)
+  if new_state == self.state then
+    states[new_state](self, line)
+
+    self.state = new_state
+  elseif self.state == 'itemize' and new_state == 'paragraph' then
+    self:append_to_itemize(line)
+  elseif self.state == 'itemize' and new_state == 'newline' then
+    table.insert(self.order, 'paragraph')
+    states[new_state](self, line)
+    self.state = new_state
+  else
+    if new_state == 'newline' then
+      table.insert(self.order, self.state)
```

Comment:
simpler maybe 

```lua
self.order[#self.order + 1] = self.state 
```

Reply:
Not a big fan of this. Also not a big fan of `#` for table length. I like to use `#` for strings and `table.getn` for tables. That will result into prettier code if you ask me.